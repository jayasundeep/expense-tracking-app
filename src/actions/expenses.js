// import { v4 as uuidv4 } from 'uuid';
import database from '../firebase/firebase';

export const addExpense = (expense = {}) => ({
    type : 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        return database.ref('expenses').push(expense)
                .then((ref) => {
                    dispatch(addExpense({
                        id : ref.key,
                        ...expense
                    }));
                });
    };
};

export const removeExpense = (id) => ({
    type : 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = (id) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove()
                .then(() => {
                    dispatch(removeExpense(id));
                });
    }
}

export const editExpense = (id, updates = {}) => 
({
    type : 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).update(updates)
                .then(() => {
                    dispatch(editExpense(id, updates));
                }).catch((e) => {
                    console.log('Error while updating...', e);
                });
    }
}

export const setExpense = (expense = {}) => ({
    type : 'SET_EXPENSE',
    expense
});

export const startSetExpense = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value')
                .then((snapshot) => {
                    const expenses = [];
                    snapshot.forEach((childSnapshot) => {
                        expenses.push({
                            id : childSnapshot.key,
                            ...childSnapshot.val()
                        });
                    });
                    dispatch(setExpense(expenses));
                });
    }
};
