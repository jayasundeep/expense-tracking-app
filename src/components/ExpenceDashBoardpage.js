import React from 'react';
import ExpenseList from './ExpensesList';

const ExpenceDashBoardpage = () => (
    <div>
        This is from ExpenceDashBoardpage!
        <ExpenseList />
    </div>
);

export default ExpenceDashBoardpage;