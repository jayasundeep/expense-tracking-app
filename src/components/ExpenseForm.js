import React from 'react';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import  moment from 'moment';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {
    constructor (props) {
        super(props);
        console.log(props.expense[0]);
        this.state = {
            description : this.props.expense ? this.props.expense[0].description : '',
            amount : this.props.expense ? this.props.expense[0].amount.toString() : '',
            note : this.props.expense ? this.props.expense[0].note : '',
            createdAt : this.props.expense ? moment(this.props.expense[0].createdAt) : moment(),
            calenderFocused : false,
            error : undefined
        }
    }
    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if(amount.match(/^\d{0,}(\.\d{0,2})?$/)){
            this.setState(() => ({ amount }))
        }
    }
    onDateChange = (createdAt) => {
        this.setState(() => ({ createdAt }))
    }
    onFocusChange = ({focused}) => {
        this.setState(() => ({ calenderFocused : focused }))
    }
    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(() => ({ error : 'You should fill both description and amount' }));
        }else {
            this.setState(() => ({ error : undefined }));
            this.props.onSubmit({
                description : this.state.description,
                amount : parseFloat(this.state.amount, 10),
                createdAt : this.state.createdAt.valueOf(),
                note : this.state.note
            });
            console.log("submitted!");
        }
    }
    render() {
        return (
            <form onSubmit={this.onSubmit}>
                <div>
                    <p>{this.state.error}</p>
                </div>
                <input
                    placeholder='Enter Description'
                    type='text'
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                >
                </input>

                <input
                    placeholder='Enter Amount'
                    type='text'
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                >
                </input>

                <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calenderFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}                   
                />

                <input
                    placeholder='Write some note(optional)'
                    type='text'
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </input>

                <button>Add Expense</button>

            </form>
        )
    }
}