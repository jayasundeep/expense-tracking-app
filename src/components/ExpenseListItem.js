import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = (props) => (
    <div>
        <h3>
            
            {props.index}. 
            <Link to={`/edit/${props.expense.id}`}>
                {props.expense.description}
            </Link>
        </h3>
        <div>
            Amount: {props.expense.amount}
            Created At: {props.expense.createdAt}
        </div>
    </div>
);

/* const mapStateToProps = (state) => {
    return {
        expenses : state.expenses
    }
} */

export default ExpenseListItem;