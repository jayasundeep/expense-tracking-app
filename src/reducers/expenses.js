const expenseReducerDefaultState = [];

const expenseReducer = (state = expenseReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense)
        case 'REMOVE_EXPENSE':
            return state.filter((item) => {return item.id !== action.id})
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if(expense.id === action.id){
                    return {
                        ...expense,
                        ...action.updates
                    }
                }else{
                    return expense;
                }
            })
        case 'GET_EXPENSE':
            return state.filter((expense) => {
                return expense.id == action.id
            })
        default :
            return state
            
    }
}

export default expenseReducer;