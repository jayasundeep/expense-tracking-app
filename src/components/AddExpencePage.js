import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import ExpenceFrom from './ExpenseForm';

export class AddExpencePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');
    }
    render(){
        return (
            <div>
                <div>
                    <h2>Add Expense Form</h2>
                </div>
                <ExpenceFrom 
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}


const mapDispatchToProps = (dispatch) => {
    return {
        addExpense : (expense) => dispatch(addExpense(expense))
    }
};

export default connect(undefined, mapDispatchToProps)(AddExpencePage);