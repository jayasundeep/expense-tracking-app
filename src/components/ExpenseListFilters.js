import React from 'react';
import { connect } from 'react-redux';
import { setFilterText, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseFilters extends React.Component {
    state = {
        calenderFocus : null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calenderFocus) => {
        this.setState(() => ({ calenderFocus }))
    };
    onTextChange = (e) => {
        this.props.setFilterText(e.target.value);
    };

    render() {
        return (
            <div>
                <input 
                    type='text' 
                    value={this.props.filters.text}
                    placeholder='Search text'
                    onChange={this.onTextChange}
                />
                <select
                    onChange={(e) => {
                        if(e.target.value === 'date'){
                            this.props.sortByDate();
                        }else{
                            this.props.sortByAmount();
                        }
                    }}
                >
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>

                <DateRangePicker 
                    focusedInput={this.state.calenderFocus}
                    numberOfMonths={1}
                    onFocusChange={this.onFocusChange}
                    onDatesChange={this.onDatesChange}
                    isOutsideRange={() => false}
                    showClearDates={true}
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    startDateId="start_date_id"
                    endDateId="end_date_id"
                />

            </div>
        );
    }
}


const mapStateToProps = (state) => ({
    filters : state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setFilterText : (text) => dispatch(setFilterText(text)), 
    sortByDate : () => dispatch(sortByDate()), 
    sortByAmount : () => dispatch(sortByAmount()), 
    setStartDate : (startDate) => dispatch(setStartDate(startDate)), 
    setEndDate : (endDate) => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseFilters);