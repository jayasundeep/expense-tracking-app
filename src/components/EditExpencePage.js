import React from 'react';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';
import { connect } from 'react-redux';

export class EditExpencePage extends React.Component {
    onSubmit = (expense) => {
        console.log(expense);
        this.props.editExpense(
            this.props.expense.id,
            expense
        );
        this.props.history.push('/')
    };
    onRemove = () => {
        console.log('I was here!');
        this.props.removeExpense(this.props.expense.id);
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                This is from EditExpencePage of {this.props.match.params.id}!
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button
                    onClick={this.onRemove}
                >
                    Remove
                </button>
            </div>
        );
    }
}


const mapStateToProps = (state, props) => {
    return {
        expense : state.expenses.filter((expense) => expense.id === props.match.params.id)[0]
    };
}

const mapDispatchToProps = (dispatch, props) => {
    return {
        editExpense : (id, expense) => dispatch(editExpense(id, expense)),
        removeExpense : (id) => dispatch((removeExpense(id)))
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpencePage);
