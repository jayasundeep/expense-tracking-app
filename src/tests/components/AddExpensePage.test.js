import React from 'react';
import { shallow } from 'enzyme';
import { AddExpencePage } from '../../components/AddExpencePage';
import expenses from '../fixtures/expenses';

let addExpense, history, wrapper;

beforeEach(() => {
    addExpense = jest.fn();
    history = { push : jest.fn() };
    wrapper = shallow(<AddExpencePage history={history} addExpense={addExpense} />);
});

test('should render add expense page correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit correctly', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(addExpense).toHaveBeenLastCalledWith(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
});
