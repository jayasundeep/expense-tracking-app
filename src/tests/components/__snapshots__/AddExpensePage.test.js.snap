// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render add expense page correctly 1`] = `
<div>
  <div>
    <h2>
      Add Expense Form
    </h2>
  </div>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
